xds/internal/xdsclient/xdsresource/cluster_resource_type.go

```diff
@@ -0,0 +1,171 @@
+//
+// Implements the Type interface.
+type clusterResourceType struct {
+}
+
+// V2TypeURL is the xDS type URL of this resource type for v2 transport.
+func (clusterResourceType) V2TypeURL() string {
+	return "type.googleapis.com/envoy.api.v2.Cluster"
+
+}
+
+// V3TypeURL is the xDS type URL of this resource type for v3 transport.
+func (clusterResourceType) V3TypeURL() string {
+	return "type.googleapis.com/envoy.config.cluster.v3.Cluster"
+}
+
+// TypeEnum identifies resources in a transport protocol agnostic way.
+func (clusterResourceType) TypeEnum() ResourceType {
+	return ClusterResource
+}
+
+// AllResourcesRequiredInSotW indicates whether this resource type requires that
+// all resources be present in every SotW response from the server. This is true
+// for a Cluster resource.
+func (clusterResourceType) AllResourcesRequiredInSotW() bool {
+	return true
+}
+
+// Decode deserializes and validates an xDS resource serialized inside the
+// provided `Any` proto, as received from the xDS management server.
+func (clusterResourceType) Decode(opts *DecodeOptions, resource *anypb.Any) (*DecodeResult, error) {
+	name, cluster, err := UnmarshalClusterResource(resource, nil, opts.Logger)
+	switch {
+	case name == "":
+		// Name is unset only when protobuf deserialization fails.
+		return nil, err
+	case err != nil:
+		// Protobuf deserialization succeeded, but resource validation failed.
+		return &DecodeResult{Name: name, Resource: &ClusterResourceData{Resource: ClusterUpdate{}}}, err
+	}
+
+	// Perform extra validation here.
+	if err := securityConfigValidator(opts.BootstrapConfig, cluster.SecurityCfg); err != nil {
+		return &DecodeResult{Name: name, Resource: &ClusterResourceData{Resource: ClusterUpdate{}}}, err
+	}
+
+	return &DecodeResult{Name: name, Resource: &ClusterResourceData{Resource: cluster}}, nil
+
+}
+
+// ClusterResourceData wraps the configuration of a Cluster resource as received
+// from the management server.
+//
+// Implements the ResourceData interface.
+type ClusterResourceData struct {
+	ResourceData
+
+	// TODO: We have always stored update structs by value. See if this can be
+	// switched to a pointer?
+	Resource ClusterUpdate
+}
+
+// Equal returns true if other is equal to r.
+func (l *ClusterResourceData) Equal(other ResourceData) bool {
+	if l == nil && other == nil {
+		return true
+	}
+	if (l == nil) != (other == nil) {
+		return false
+	}
+	return proto.Equal(l.Resource.Raw, other.Raw())
+
+}
+
+// ToJSON returns a JSON string representation of the resource data.
+func (l *ClusterResourceData) ToJSON() string {
+	return pretty.ToJSON(l.Resource)
+}
+
+// Raw returns the underlying raw protobuf form of the cluster resource.
+func (l *ClusterResourceData) Raw() *anypb.Any {
+	return l.Resource.Raw
+}
+
+// ClusterResourceWatcher wraps the callbacks invoked by the xDS client
+// implementation for different events corresponding to the listener resource
+// being watched.
+type ClusterResourceWatcher interface {
```

Comment:
If we had generics, we would only need one of these, right?  Instead of one per resource?  So maybe TODO?

Reply:
Even with generics, we would need one of these for each resource type. The reason for that is to ensure that the `OnResourceChanged()` method returns a concrete implementation of the `ResourceData` interface, thereby ensuring type safety and not leaving it to the caller to perform a type assertion.